File: python/day9.py
import time


def solve_part1(input):
    data = decompress(input, False)
    j = 0
    for i in range(len(data) - 1, 1, -1):
        j = data.index(".", j)
        if j > i:
            break
        data[j] = data[i]
        data[i] = "."

    checksum = 0
    for i, n in enumerate(data):
        if n == ".":
            break
        checksum += i * int(n)
    return checksum


def solve_part2(input):
    data = decompress(input, True)
    for i in range(len(data) - 1, 0, -1):
        id, n = data[i]
        if id == ".":
            continue
        for j, (_, space) in enumerate(data):
            if j >= i:
                break
            if data[j][0] != ".":
                continue
            if n == space:
                data[i], data[j] = data[j], data[i]
                break
            if n < space:
                data[j][1] -= n
                data[i][0] = "."
                data.insert(j, [id, n])
                break

    return checksum(data)

def decompress(input, as_blocks):
    result = []
    data = input[::2]
    free = input[1::2]
    i = j = 0
    while i < len(data) and j < len(free):
        if as_blocks:
            result.extend([[str(i), data[i]], [".", free[j]]])
        else:
            result.extend([str(i)] * data[i] + ["."] * free[j])
        i += 1
        j += 1
    for n in range(i, len(data)):
        if as_blocks:
            result.append([str(n), data[n]])
        else:
            result.extend([str(n)] * data[n])
    for n in range(j, len(free)):
        if as_blocks:
            result.append([".", free[n]])
        else:
            result.extend(["."] * free[n])
    return result

def checksum(data):
    checksum = 0
    i = 0
    for id, n in data:
        if id == ".":
            i += n
            continue
        for _ in range(n):
            checksum += i * int(id)
            i += 1
    return checksum

if __name__ == "__main__":
    with open("day9.input", "r") as f:
        input = f.read()

    input = [int(n) for n in list(input.strip())]

    start = time.perf_counter()
    result = solve_part1(input)
    end = time.perf_counter()
    print(f"Part 1: {result} => {end-start:.6f} seconds")

    start = time.perf_counter()
    result = solve_part2(input)
    end = time.perf_counter()
    print(f"Part 2: {result} => {end-start:.6f} seconds")

File: zig/day9.zig
const std = @import("std");

const Block = struct { []const u8, usize };

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const allocator = gpa.allocator();

    const input = try std.fs.cwd().readFileAlloc(allocator, "day9.input", std.math.maxInt(usize));
    const trimmed = std.mem.trim(u8, input, " \n");

    var timer = try std.time.Timer.start();
    var result = try solvePart1(allocator, trimmed);
    var elapsed_ns = timer.read();
    var elapsed = @as(f64, @floatFromInt(elapsed_ns)) / std.time.ns_per_s;
    std.debug.print("Part 1: {d:<3}  =>  {d:.6} seconds\n", .{ result, elapsed });

    timer.reset();
    result = try solvePart2(allocator, trimmed);
    elapsed_ns = timer.read();
    elapsed = @as(f64, @floatFromInt(elapsed_ns)) / std.time.ns_per_s;
    std.debug.print("Part 2: {d:<3}  =>  {d:.6} seconds\n", .{ result, elapsed });
}

fn solvePart1(allocator: std.mem.Allocator, input: []const u8) !usize {
    var data = try decompress(allocator, input, []const u8);
    var j: usize = 0;
    for (0..data.items.len) |idx| {
        const i = data.items.len - 1 - idx;
        j = blk: for (j.., data.items[j..]) |k, s| {
            if (s[0] == '.') break :blk k;
        } else break;
        if (j > i) break;
        data.items[j] = data.items[i];
        data.items[i] = ".";
    }

    var result: usize = 0;
    for (0.., data.items) |i, n| {
        if (n[0] == '.') break;
        result += i * try std.fmt.parseInt(usize, n, 10);
    }
    return result;
}

fn solvePart2(allocator: std.mem.Allocator, input: []const u8) !usize {
    var data = try decompress(allocator, input, Block);
    for (0..data.items.len) |idx| {
        const i = data.items.len - 1 - idx;
        const id, const n = data.items[i];
        if (id[0] == '.') continue;
        for (0.., data.items) |j, d| {
            _, const space = d;
            if (j >= i) break;
            if (data.items[j][0][0] != '.') continue;
            if (n == space) {
                std.mem.swap(Block, &data.items[i], &data.items[j]);
                break;
            }
            if (n < space) {
                data.items[j][1] -= n;
                data.items[i][0] = ".";
                try data.insert(j, .{ id, n });
                break;
            }
        }
    }
    return try checksum(data.items);
}

fn decompress(allocator: std.mem.Allocator, input: []const u8, T: type) !std.ArrayList(T) {
    var result = std.ArrayList(T).init(allocator);

    for (0.., input) |i, num| {
        const value = if (i % 2 == 0) try std.fmt.allocPrint(allocator, "{d}", .{i / 2}) else ".";
        switch (T) {
            Block => {
                try result.append(.{ value, num - '0' });
            },
            []const u8 => {
                for (0..num - '0') |_| {
                    try result.append(value);
                }
            },
            else => unreachable,
        }
    }

    return result;
}

fn checksum(data: []const Block) !usize {
    var result: usize = 0;
    var i: usize = 0;
    for (data) |block| {
        const id, const n = block;
        if (id[0] == '.') {
            i += n;
            continue;
        }
        for (0..n) |_| {
            result += i * try std.fmt.parseInt(usize, id, 10);
            i += 1;
        }
    }
    return result;
}
